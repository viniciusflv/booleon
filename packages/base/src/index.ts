export { View } from './View';
export { Text } from './Text';
