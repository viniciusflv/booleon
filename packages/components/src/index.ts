export * from './Button';
export * from './Field';
