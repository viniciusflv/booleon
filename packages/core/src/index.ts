export * from "./Base";
export * from "./Text";
