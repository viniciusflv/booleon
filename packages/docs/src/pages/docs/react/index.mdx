---
icon: react
title: Booleon React
---

import { logo } from '../../../assets';
import { Svg } from '@booleon/ui';

<div align="center">
  <Svg w_144 h_144 fl_sd_2 dark__fl_sd_4 {...logo} />
  <h1 align="center">booleon/core</h1>
  <p align="center" style="font-size: 1.2rem;">
    Highly typed declarative CSS-in-JS
  </p>
  <hr />
</div>

Booleon React is a `DX` focused `library` for, a highly typed `declarative`, styling, using component `boolean` properties. Works by mapping a `pre-defined` entry of css with keys that can be static (`string`) or dynamic (`RegEx`).

![NPM version](https://badgen.net/npm/v/@booleon/react)
![NPM version](https://badgen.net/github/commits/viniciusflv/booleon)
![NPM version](https://badgen.net//packagephobia/publish/@booleon/react)
![NPM version](https://badgen.net/bundlephobia/min/@booleon/react)
![NPM version](https://badgen.net/bundlephobia/minzip/@booleon/react)

# Get Started

## Instalation

```shell
npm i @booleon/react
```

```shell
yarn add @booleon/react
```

## Create Component

```jsx
import { booleon } from '@booleon/react';
import { background } from '@booleon/module';

export const ButtonStyled = booleon.button(background);
```

## Basic Usage

```jsx
export const MyComponent = () => {
  return (
    <ButtonStyled bg_color_f00 />
  )
}
```

## Conditional styling

Only applies css if a `condition` is satisfied.

```jsx
export const MyComponent = ({ color }) => {
  return (
    <ButtonStyled bg_color_f00={color === 'red'} />
  )
}
```
## Responsiveness

Add styles to `media queries`.

```jsx
export const MyComponent = () => {
  return (
    <ButtonStyled md__bg_color_f00 />
  )
}
```

## Pseudo Elements

Applies, possible nested, `pseudo elements`.

```jsx
export const MyComponent = () => {
  return (
    <ButtonStyled hover__bg_color_f00 />
  )
}
```

## Keyframes

Create simple and fast `keyframes`.

```jsx
import { booleon } from '@booleon/react';
import { container, animation } from '@booleon/module';

const Component = booleon.div(container, animation);
export const MyComponent = () => {
  return (
    <Component
      kf__myAnimation
      from__top
      to__top_full
      ani_name_myAnimation
      ani_iteration_infinite
    />
  )
}
```

## Dark mode

Allows `dark theme` props declarations. (default from `prefers-color-scheme`)

```jsx
import { useTheme } from '@booleon/react';

export const MyComponent = () => {
  const { toggleTheme } = useTheme();
  return (
    <ButtonStyled dark__bg_color_f00 onClick={toggleTheme} />
  )
}
```
