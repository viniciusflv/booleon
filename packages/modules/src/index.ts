export * from './constants';
export * from './css';
export * from './types';
