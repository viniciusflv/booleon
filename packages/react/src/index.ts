export * from './hocBooleon';
export * from './useBooleon';
