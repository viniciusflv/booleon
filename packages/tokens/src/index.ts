export default {
  colors: {
    brand: {
      primary: 'hsl(25, 99%, 42%)',
      secondary: 'hsl(25, 100%, 21%)',
    },
    base: {
      emerald: {
        '065': 'hsl(135, 94%, 65%)',
        '040': 'hsl(175, 59%, 40%)',
        '030': 'hsl(68, 100%, 30%)',
        '014': 'hsl(192, 81%, 14%)',
      },
      citrine: {
        '071': 'hsl(31, 100%, 71%)',
        '044': 'hsl(18, 80%, 44%)',
        '042': 'hsl(25, 99%, 42%)',
        '021': 'hsl(25, 100%, 21%)',
      },
      amber: {
        '076': 'hsl(65, 92%, 76%)',
        '035': 'hsl(45, 100%, 35%)',
      },
      sapphire: {
        '077': 'hsl(191, 97%, 77%)',
        '049': 'hsl(205, 69%, 49%)',
      },
      amethyst: {
        '078': 'hsl(265, 89%, 78%)',
        '074': 'hsl(326, 100%, 74%)',
        '051': 'hsl(225, 27%, 51%)',
        '031': 'hsl(232, 14%, 31%)',
        '018': 'hsl(231, 15%, 18%)',
        '007': 'hsl(232, 23%, 7%)',
      },
      granite: {
        '100': 'hsl(0, 0%, 100%)',
        '098': 'hsl(20, 82%, 98%)',
        '096': 'hsl(60, 30%, 96%)',
        '084': 'hsl(0, 0%, 84%)',
        '063': 'hsl(210, 6%, 63%)',
        '060': 'hsl(180, 7%, 60%)',
        '045': 'hsl(196, 13%, 45%)',
        '040': 'hsl(194, 14%, 40%)',
        '00080': 'hsl(0, 0%, 0%, .8)',
        '000': 'hsl(0, 0%, 0%)',
      },
    },
    dark: {
      code: {
        background: 'hsl(231, 15%, 18%)',
        selection: 'hsl(232, 14%, 31%)',
        baseColor: 'hsl(60, 30%, 96%)',
        comment: 'hsl(225, 27%, 51%)',
        punctuation: 'hsl(60, 30%, 96%)',
        operator: 'hsl(326, 100%, 74%)',
        parameter: 'hsl(31, 100%, 71%)',
        property: 'hsl(265, 89%, 78%)',
        string: 'hsl(65, 92%, 76%)',
        function: 'hsl(135, 94%, 65%)',
        builtin: 'hsl(191, 97%, 77%)',
        keyword: 'hsl(326, 100%, 74%)',
        variable: 'hsl(265, 89%, 78%)',
      },
    },
    light: {
      code: {
        background: 'hsl(20, 82%, 98%)',
        selection: 'hsl(192, 81%, 14%)',
        baseColor: 'hsl(196, 13%, 45%)',
        comment: 'hsl(180, 7%, 60%)',
        punctuation: 'hsl(194, 14%, 40%)',
        operator: 'hsl(194, 14%, 40%)',
        parameter: 'hsl(18, 80%, 44%)',
        property: 'hsl(205, 69%, 49%)',
        string: 'hsl(175, 59%, 40%)',
        function: 'hsl(45, 100%, 35%)',
        builtin: 'hsl(205, 69%, 49%)',
        keyword: 'hsl(68, 100%, 30%)',
        variable: 'hsl(18, 80%, 44%)',
      },
    },
  },
  sizes: {
    base: {
      none: '0rem',
      px: '0.0625rem',
      xs: '0.125rem',
      sm: '0.25rem',
      md: '0.375rem',
      lg: '0.5rem',
      xl: '0.75rem',
      '2xl': '1rem',
      '3xl': '1.5rem',
      '4xl': '2rem',
      '5xl': '3rem',
      '6xl': '4rem',
      '7xl': '6rem',
      '8xl': '8rem',
      '9xl': '16rem',
      '10xl': '12rem',
      '11xl': '24rem',
      '12xl': '32rem',
      '13xl': '48rem',
      '14xl': '64rem',
      '15xl': '96rem',
      '16xl': '128rem',
    },
    border: {
      width: {
        none: '0rem',
        px: '0.0625rem',
        xs: '0.125rem',
        sm: '0.25rem',
        md: '0.375rem',
        lg: '0.5rem',
        xl: '0.75rem',
      },
      radius: {
        none: '0rem',
        px: '0.0625rem',
        xs: '0.125rem',
        sm: '0.25rem',
        md: '0.375rem',
        lg: '0.5rem',
        xl: '0.75rem',
        circle: '0.75rem',
        pill: '0.75rem',
      },
    },
    relative: {
      '0': '10',
      '1': '20',
      '2': '30',
      '3': '40',
      '4': '50',
      '5': '60',
      '6': '70',
      '7': '80',
      '8': '90',
      '9': '100',
    },
    percentage: {
      '10': '10%',
      '20': '20%',
      '30': '30%',
      '40': '40%',
      '50': '50%',
      '60': '60%',
      '70': '70%',
      '80': '80%',
      '90': '90%',
      '100': '100%',
    },
    screen: {
      width: {
        '10': '10vw',
        '20': '20vw',
        '30': '30vw',
        '40': '40vw',
        '50': '50vw',
        '60': '60vw',
        '70': '70vw',
        '80': '80vw',
        '90': '90vw',
        '100': '100vw',
      },
      height: {
        '10': '10vh',
        '20': '20vh',
        '30': '30vh',
        '40': '40vh',
        '50': '50vh',
        '60': '60vh',
        '70': '70vh',
        '80': '80vh',
        '90': '90vh',
        '100': '100vh',
      },
    },
    sizing: {
      none: '0rem',
      px: '0.0625rem',
      xs: '0.125rem',
      sm: '0.25rem',
      md: '0.375rem',
      lg: '0.5rem',
      xl: '0.75rem',
      '2xl': '1rem',
      '3xl': '1.5rem',
      '4xl': '2rem',
      '5xl': '3rem',
      '6xl': '4rem',
      '7xl': '6rem',
      '8xl': '8rem',
      '9xl': '16rem',
      '10xl': '12rem',
      '11xl': '24rem',
      '12xl': '32rem',
      '13xl': '48rem',
      '14xl': '64rem',
      '15xl': '96rem',
      '16xl': '128rem',
    },
    spacing: {
      none: '0rem',
      px: '0.0625rem',
      xs: '0.125rem',
      sm: '0.25rem',
      md: '0.375rem',
      lg: '0.5rem',
      xl: '0.75rem',
    },
  },
} as const;
