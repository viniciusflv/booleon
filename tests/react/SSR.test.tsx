/**
 * @jest-environment node
 */
import React from 'react';
import ReactDOM from 'react-dom/server';

import { booleon } from '../../packages/react/src';

describe('SSR', () => {
  test('renderToString', () => {
    const MyComponent = booleon.section([
      ['flex', () => 'display:flex;'],
    ] as const);
    const string = ReactDOM.renderToString(<MyComponent flex />);
    expect(string).toMatchSnapshot();
  });

  test('renderToStaticMarkup', () => {
    const MyComponent = booleon.section([
      ['flex', () => 'display:flex;'],
    ] as const);
    const string = ReactDOM.renderToStaticMarkup(<MyComponent flex />);
    expect(string).toMatchSnapshot();
  });

  test('renderToNodeStream', async (done) => {
    const MyComponent = booleon.section([
      ['flex', () => 'display:flex;'],
    ] as const);
    const stream = ReactDOM.renderToNodeStream(<MyComponent flex />);
    const string = await new Promise((resolve, reject) => {
      let received = '';
      stream.on('data', (chunk) => (received += chunk));
      stream.on('end', () => resolve(received));
      stream.on('error', reject);
    });

    expect(string).toMatchSnapshot();

    done();
  });

  test('renderToStaticNodeStream', async (done) => {
    const MyComponent = booleon.section([
      ['flex', () => 'display:flex;'],
    ] as const);
    const staticStream = ReactDOM.renderToStaticNodeStream(
      <MyComponent flex />,
    );
    const string = await new Promise((resolve, reject) => {
      let received = '';
      staticStream.on('data', (chunk) => (received += chunk));
      staticStream.on('end', () => resolve(received));
      staticStream.on('error', reject);
    });

    expect(string).toMatchSnapshot();

    done();
  });
});
