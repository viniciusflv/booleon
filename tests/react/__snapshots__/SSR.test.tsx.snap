// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SSR renderToNodeStream 1`] = `"<style data-booleon=\\"bl-1980822999\\">.bl-1980822999{display:flex;}</style><section class=\\"bl-1980822999\\"></section>"`;

exports[`SSR renderToStaticMarkup 1`] = `"<style data-booleon=\\"bl-1980822999\\">.bl-1980822999{display:flex;}</style><section class=\\"bl-1980822999\\"></section>"`;

exports[`SSR renderToStaticNodeStream 1`] = `"<style data-booleon=\\"bl-1980822999\\">.bl-1980822999{display:flex;}</style><section class=\\"bl-1980822999\\"></section>"`;

exports[`SSR renderToString 1`] = `"<style data-booleon=\\"bl-1980822999\\">.bl-1980822999{display:flex;}</style><section class=\\"bl-1980822999\\"></section>"`;
